applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/eigenfrequency_response_function_utility.h

```diff
@@ -64,327 +64,351 @@ namespace Kratos
+        }
+
+        // Element-wise computation of gradients
+        for (auto& elem_i : mrModelPart.Elements())
+        {
+            Matrix mass_matrix_org;
+            Matrix LHS_org;
+            Vector dummy;
+            Matrix aux_matrix = Matrix(0,0);
+            Vector aux_vector = Vector(0);
+            elem_i.CalculateMassMatrix(mass_matrix_org, CurrentProcessInfo);
+            elem_i.CalculateLocalSystem(LHS_org, dummy ,CurrentProcessInfo);
+
+            // Predetermine the necessary eigenvectors
+            int num_dofs_element = mass_matrix_org.size1();
+            std::vector<Vector> eigenvectors_of_element(num_of_traced_eigenfrequencies,Vector(0));
+            for(std::size_t i = 0; i < num_of_traced_eigenfrequencies; i++)
+                DetermineEigenvectorOfElement(elem_i, mTracedEigenfrequencyIds[i], num_dofs_element, eigenvectors_of_element[i]);
+
+            // Computation of derivative of state equation w.r.t. node coordinates
+            for(auto& node_i : elem_i.GetGeometry())
+            {
+                Vector gradient_contribution(3, 0.0);
+                Matrix perturbed_LHS = Matrix(0,0);
+                Matrix perturbed_mass_matrix = Matrix(0,0);
+
+                for(std::size_t coord_dir_i = 0; coord_dir_i < CurrentProcessInfo.GetValue(DOMAIN_SIZE); coord_dir_i++)
+                {
+                    node_i.GetInitialPosition()[coord_dir_i] += mDelta;
+
+                    elem_i.CalculateMassMatrix(perturbed_mass_matrix, CurrentProcessInfo);
+                    perturbed_mass_matrix = (perturbed_mass_matrix - mass_matrix_org) / mDelta;
+
+                    elem_i.CalculateLocalSystem(perturbed_LHS, dummy ,CurrentProcessInfo);
+                    perturbed_LHS = (perturbed_LHS - LHS_org) / mDelta;
+
+                    for(std::size_t i = 0; i < num_of_traced_eigenfrequencies; i++)
+                    {
+                        aux_matrix.clear();
+                        aux_vector.clear();
+
+                        aux_matrix = perturbed_LHS;
+                        aux_matrix -= (perturbed_mass_matrix * traced_eigenvalues[i]);
+                        aux_vector = prod(aux_matrix , eigenvectors_of_element[i]);
+
+                        gradient_contribution[coord_dir_i] += gradient_prefactors[i] * inner_prod(eigenvectors_of_element[i] , aux_vector) * mWeightingFactors[i];
+                    }
+
+                   node_i.GetInitialPosition()[coord_dir_i] -= mDelta;
+                }
+                noalias(node_i.FastGetSolutionStepValue(SHAPE_SENSITIVITY)) += gradient_contribution;
+            }
+        }
+    }
+
+    // --------------------------------------------------------------------------
+    double GetEigenvalue(int eigenfrequency_id)
+    {
+        const int num_of_computed_eigenvalues = (mrModelPart.GetProcessInfo()[EIGENVALUE_VECTOR]).size();
+        return (mrModelPart.GetProcessInfo()[EIGENVALUE_VECTOR])[eigenfrequency_id-1];
```

Comment:
You could add a `KRATOS_ERROR_IF` if the requested id is larger than what is available (then the other variable wouldn't be unused)